data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,498 @@
+}, {
+    "type": "date",
+    "name": "deactivated_date",
+    "mode": "required",
+    "description": "The deactivation date to base retractions on"
+}]
+
+
+def add_console_logging(add_handler):
+    """
+    This config should be done in a separate module, but that can wait
+    until later.  Useful for debugging.
+    """
+    try:
+        os.makedirs(LOGS_PATH)
+    except OSError:
+        # directory already exists.  move on.
+        pass
+
+    name = datetime.now().strftime(
+        os.path.join(LOGS_PATH, 'ehr_deactivated_retraction-%Y-%m-%d.log'))
+    logging.basicConfig(filename=name,
+                        level=logging.INFO,
+                        format='{asctime} - {name} - {levelname} - {message}',
+                        style='{')
+
+    if add_handler:
+        handler = logging.StreamHandler()
+        handler.setLevel(logging.INFO)
+        formatter = logging.Formatter('{levelname} - {name} - {message}',
+                                      style='{')
+        handler.setFormatter(formatter)
+        logging.getLogger('').addHandler(handler)
+
+
+def get_pids_table_info(project_id, dataset_id):
+    """
+    This function gets all table information for the dataset and filters to only retain tables with the field person_id
+
+    :param project_id: bq name of project_id
+    :param dataset_id: ba name of dataset_id
+    :return: dataframe with table information schema from BigQuery
+    """
+    all_table_info_query = TABLE_INFORMATION_SCHEMA.render(project=project_id,
+                                                           dataset=dataset_id)
+    result_df = bq.query(q=all_table_info_query, project_id=project_id)
+    # Get list of tables that contain person_id
+    pids_tables = []
+    for i, row in result_df.iterrows():
+        column = getattr(row, 'column_name')
+        table = getattr(row, 'table_name')
+        if 'person_id' in column:
+            pids_tables.append(table)
+    pids_tables_info_df = result_df[result_df['table_name'].isin(pids_tables)]
+    return pids_tables_info_df
+
+
+def get_date_info_for_pids_tables(project_id):
+    """
+    Loop through tables within all datasets and determine if the table has an end_date date or a date field. Filtering
+    out person and death table and keeping only tables with PID and an upload or start/end date associated.
+
+    :param project_id: bq name of project_id
+    :return: filtered dataframe which includes the following columns for each table in each dataset with a person_id
+    'project_id', 'dataset_id', 'table', 'date_column', 'start_date_column', 'end_date_column'
+    """
+    # Create empty df to append to for final output
+    date_fields_info_df = pd.DataFrame()
+
+    # Loop through datasets
+    LOGGER.info(
+        "Looping through datasets to filter and create dataframe with correct date field to determine retraction"
+    )
+
+    dataset_obj = bq.list_datasets(project_id)
```

Comment:
agree, should this be a separate ticket or should I just make this adjustment with this PR?

Reply:
This would be part of this PR, because the example code is assuming it is placed in this module.  You just wouldn't be calling the `bq.list_datasets(project_id)` function anymore.

Reply:
perfect, I will make this adjustment.

Reply:
I was getting an error when trying to change the global variable CLIENT, so I have updated `bq.py` to the following, let me know if that makes sense. I also left in the logic that was there before for `project_id`.

```
CLIENT = None

def get_client(project_id=None):
    """
    Ensure only one client is created and reused

    :param project_id:  project to get a client for
    :returns: a big query client object
    """
    if CLIENT:
        return CLIENT
    if project_id is None:
        LOGGER.info(
            'You should specify project_id for a reliable experience.'
            'Defaulting to %s.', os.environ.get(PROJECT_ID))
        return bigquery.Client()
    return bigquery.Client(project=project_id)
```

Reply:
The above never sets the `CLIENT` global object.  Should it?  My original code example is flawed.  I tested this example more thoroughly.  This will work to define a `CLIENT` variable in the retraction module.
```
from google.cloud import bigquery

# global variable declaration
CLIENT = None

def get_client(project_id):
    """                 
    Ensure only one client is created and reused
                        
    :param project_id:  project to get a client for
    :returns: a big query client object
    """ 
    # declaring that this function should use the global CLIENT variable instead of 
    # creating a local variable.  This local variable was causing the first failure.
    global CLIENT       
    if not CLIENT:
        CLIENT = bigquery.Client(project=project_id)

    return CLIENT
```

Reply:
this has been adjusted per our phone conversation!